File: app/src/main/java/com/mirash/passkeeper/activity/main/MainActivityModel.java
package com.mirash.passkeeper.activity.main;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import com.mirash.passkeeper.db.Credentials;
import com.mirash.passkeeper.db.RepositoryProvider;
import com.mirash.passkeeper.model.CredentialsItem;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executors;

/**
 * @author Mirash
 */
public class MainActivityModel extends AndroidViewModel {
    private final LiveData<List<Credentials>> credentialsModelLiveData;

    public MainActivityModel(@NonNull Application application) {
        super(application);
        credentialsModelLiveData = RepositoryProvider.getCredentialsRepository().getAllCredentialsLiveData();
    }

    public LiveData<List<Credentials>> getCredentialsModelLiveData() {
        return credentialsModelLiveData;
    }

    public void handleOrderChanged(List<CredentialsItem> items) {
        Executors.newSingleThreadScheduledExecutor().execute(() -> {
            handleOrderChangedSync(items);
        });
    }

    public void handleOrderChangedSync(List<CredentialsItem> items) {
        List<Credentials> credentials = new ArrayList<>();
        for (int i = 0; i < items.size(); i++) {
            int id = items.get(i).getId();
            Credentials credById = RepositoryProvider.getCredentialsRepository().getCredentialsByIdSync(id);
            credById.setPosition(i);
            credentials.add(credById);
        }
        RepositoryProvider.getCredentialsRepository().updateAll(credentials);
    }
}

File: app/src/main/java/com/mirash/familiar/activity/main/MainActivityModel.kt
package com.mirash.familiar.activity.main

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import com.mirash.familiar.db.Credentials
import com.mirash.familiar.db.RepositoryProvider
import com.mirash.familiar.db.User
import com.mirash.familiar.eventmanager.EventManager
import com.mirash.familiar.eventmanager.event.DataEvent
import com.mirash.familiar.eventmanager.subscribe
import com.mirash.familiar.eventmanager.subscription.Subscription
import com.mirash.familiar.model.credentials.CredentialsItem
import com.mirash.familiar.tool.AppEventType
import com.mirash.familiar.user.TAG_USER
import com.mirash.familiar.user.UserControl
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking

/**
 * @author Mirash
 */
class MainActivityModel(application: Application, private var callback: MainModelCallback? = null) :
    AndroidViewModel(application) {

    private var userLiveData: LiveData<User> = RepositoryProvider.userRepository.getById(UserControl.userId)
    private var credentialsLiveData: LiveData<List<Credentials>> =
        RepositoryProvider.credentialsRepository.getAllByUserId(UserControl.userId)

    init {
        callback?.setUserObservers(userLiveData, credentialsLiveData)
    }

    val usersLiveData: LiveData<List<User>> = RepositoryProvider.userRepository.getAll()


    private val userSubscription: Subscription = EventManager.subscribe<DataEvent<Long>>(AppEventType.USER_SET) {
        Log.d(TAG_USER, "MainActivityModel USER_SET event: ${it.data}")
        callback?.clearUserObservers(userLiveData, credentialsLiveData)
        userLiveData = RepositoryProvider.userRepository.getById(it.data)
        credentialsLiveData = RepositoryProvider.credentialsRepository.getAllByUserId(it.data)
        callback?.setUserObservers(userLiveData, credentialsLiveData)
    }

    fun handleOrderChanged(items: List<CredentialsItem>) {
        runBlocking { launch(Dispatchers.Default) { handleOrderChangedSync(items) } }
    }

    private fun handleOrderChangedSync(items: List<CredentialsItem>) {
        val credentials: MutableList<Credentials> = ArrayList()
        for (i in items.indices) {
            val id = items[i].id
            RepositoryProvider.credentialsRepository.getByIdSync(id)?.let {
                it.position = i
                credentials.add(it)
            }
        }
        RepositoryProvider.credentialsRepository.updateAll(credentials)
    }

    override fun onCleared() {
        super.onCleared()
        callback = null
        userSubscription.unsubscribe()
    }
}
